src/cdsetool/download.py

```diff
@@ -10,33 +10,76 @@
 import tempfile
 import time
 import shutil
+
+from dataclasses import dataclass
+
 from cdsetool._processing import _concurrent_process
 from cdsetool.credentials import Credentials
 from cdsetool.logger import NoopLogger
 from cdsetool.monitor import NoopMonitor
 
 
+@dataclass
+class DownloadResult:
+    """
+    Abstract base class for download results
+    """
+
+
+@dataclass
+class DownloadSuccess(DownloadResult):
+    """
+    Result class for successful downloads
+    """
+
+    def __init__(self, feature, filename):
+        self.feature = feature
+        self.filename = filename
+
+    def __str__(self):
+        return f"Downloaded {self.feature.get('id')} to {self.filename}"
+
+
+@dataclass
+class DownloadFailure(DownloadResult):
+    """
+    Result class for failed downloads
+    """
+
+    def __init__(self, feature, message):
+        self.feature = feature
```

Comment:
Can `feature` be in the base class? And the class only uses the id part of the feature, so perhaps id should be stored directly?

Reply:
I agree we should store feature in the base class, but my intention was to allow the user full access to the feature object.